langchain4j-core/src/main/java/dev/langchain4j/rag/RetrievalAugmentor.java

```diff
@@ -23,5 +23,5 @@ public interface RetrievalAugmentor {
      * @param metadata    The {@link Metadata} that may be useful or necessary for retrieval and augmentation.
      * @return The augmented {@link UserMessage}.
      */
-    UserMessage augment(UserMessage userMessage, Metadata metadata);
+    T augment(UserMessage userMessage, Metadata metadata);
```

Comment:
This is a breaking change, is there a way to avoid this?

Reply:
The purpose of this is to retrieve the contents used for augmentation.

I have two solutions in mind:

1. **First solution:** Add another method in the `RetrievalAugmentor` which returns another object containing the `UserMessage` and the list of contents.
2. **Second solution:** Add the list of contents (`List<TextSegments>`) in the `UserMessage` class.

I noticed another list of contents in the `UserMessage`, but it's only used to specify the type of content used for augmentation.

What do you suggest to make retrieving the `UserMessage` with the list of content possible in the `DefaultAiServices` class?

Reply:
I assume that in the long term, regardless of this change, we will have to make internals of `UserMessage` more fine-grained, e.g. having user message and retrieved contents separate (solution 2). I foresee that LLM APIs in the future will separate user message from supporting (retrieved) content to be able to adjust/fine-tune their LLMs better. Some LLM providers [already do this](https://docs.cohere.com/reference/chat) (see `documents` parameter). But this is pretty big change, and introduces more complexity, so I would better postpone it untill it will be really required.

I guess for now we can go with the first solution, something like this?

```java 
public interface RetrievalAugmentor {

    @Deprecated // let's deprecate this
    UserMessage augment(UserMessage userMessage, Metadata metadata); // existing API

    default AugmentationResult augment(AugmentationRequest augmentationRequest) { // new API
        UserMessage augmented = augment(augmentationRequest.chatMessage, augmentationRequest.metadata);
        return new AugmentationResult(augmented, null);
    }
}

// introduce request/result classes to make it more extendable in the future
public class AugmentationRequest {

    ChatMessage chatMessage; // instead of UserMessage, this will be useful for issue #793
    Metadata metadata;
}

public class AugmentationResult {

    ChatMessage augmentedChatMessage;
    List<Content> retrievedContents;
}
```

This feature will not work for existing custom implementations of `RetrievalAugmentor`, but I guess it is fine as there are most probably not many such users and this class is explicitly marked `@Experimental`.

Most users are probably using `DefaultRetrievalAugmentor`, where we will have a proper implementation of the new method.

WDYT?

Reply:
I'll include the ChatMessage in a separate PR since it requires numerous modifications

Reply:
@KaisNeffati what is the change exactly? Is it breaking?

Reply:
To utilize `ChatMessage` within both `AugmentationRequest` and `AugmentationResult`, the interface should be adjusted as follows:
```

    @Deprecated
    UserMessage augment(UserMessage userMessage, Metadata metadata);

    UserMessage augment(ChatMessage chatMessage, Metadata metadata);

    /**
     * Augments the provided {@link AugmentationRequest} with retrieved content.
     *
     * @param augmentationRequest The {@link AugmentationRequest} containing the user message and metadata.
     * @return The {@link AugmentationResult} containing the augmented user message.
     */
    default AugmentationResult augment(AugmentationRequest augmentationRequest) { // new API
        UserMessage augmented = augment(augmentationRequest.getChatMessage(), augmentationRequest.getMetadata());
        return AugmentationResult.builder()
                .augmentedUserMessage(augmented)
                .build();
    }
```

This adjustment necessitates modifying the contentInjector to accept:

`    UserMessage inject(List<Content> contents, ChatMessage chatMessage);
`
etc...

So for this PR i kept using the UserMessage , And i'll include the use of ChatMessage in another PR